plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/loops/routing/OrthogonalSelfLoopRouter.java

```diff
@@ -0,0 +1,394 @@
+        computePositions(positions, slHolder, PortSide.SOUTH);
+        computePositions(positions, slHolder, PortSide.WEST);
+        
+        return positions;
+    }
+
+    /**
+     * Initializes the entry for each routing slot on the north or south sides with the maximum height of labels it
+     * houses.
+     */
+    private void initializeWithMaxLabelHeight(final double[][] positions, final SelfLoopHolder slHolder,
+            final PortSide portSide) {
+        
+        assert portSide == PortSide.NORTH || portSide == PortSide.SOUTH;
+        
+        double[] sidePositions = positions[portSide.ordinal()];
+        
+        for (SelfHyperLoop slLoop : slHolder.getSLHyperLoops()) {
+            SelfHyperLoopLabels slLabels = slLoop.getSLLabels();
+            if (slLabels != null && slLabels.getSide() == portSide) {
+                int routingSlot = slLoop.getRoutingSlot(portSide);
+                sidePositions[routingSlot] = Math.max(sidePositions[routingSlot], slLabels.getSize().y);
+            }
+        }
+    }
+    
+    /**
+     * Computes the positions for each routing slot, taking labels into account.
+     */
+    private void computePositions(final double[][] positions, final SelfLoopHolder slHolder, final PortSide portSide) {
+        double currPos = computeBaselinePosition(slHolder, portSide);
+        
+        // For northern and western coordinates, we have to subtract from the current position
+        double factor = portSide == PortSide.NORTH || portSide == PortSide.WEST ? -1 : 1;
+        
+        double[] sidePositions = positions[portSide.ordinal()];
+        for (int slot = 0; slot < sidePositions.length; slot++) {
+            // The slot entry currently contains the height or width of the largest label in that slot
+            double largestLabelSize = sidePositions[slot];
+            if (largestLabelSize > 0) {
+                // Account for label spacing
+                largestLabelSize += LABEL_EDGE_DISTANCE;
+            }
+            
+            // Place the slot at the current position and advance the position
+            sidePositions[slot] = currPos;
+            currPos += factor * (largestLabelSize + DISTANCE);
+        }
+    }
+
+    /**
+     * Based on the margin area, this computes the offset from the node origin to add to escape the area occupied by
+     * ports.
+     */
+    private double computeBaselinePosition(final SelfLoopHolder slHolder, final PortSide portSide) {
+        LNode lNode = slHolder.getLNode();
+        LMargin lMargins = lNode.getMargin();
+        
+        switch (portSide) {
+        case NORTH:
+            return -lMargins.top - DISTANCE;
+        case EAST:
+            return lNode.getSize().x + lMargins.right + DISTANCE;
+        case SOUTH:
+            return lNode.getSize().y + lMargins.bottom + DISTANCE;
+        case WEST:
+            return -lMargins.left - DISTANCE;
+        default:
+            assert false;
+            return -1;
```

Comment:
🤔

Reply:
If we run into this case, it's an internal bug, which explains the `assert` instead of an exception. It doesn't really matter, then, what we return, since things already are very broken. ;)